### PR TITLE
Add checking for no_mangle to unsafe_code lint

### LINKED ISSUE #72188: Add an allow-by-default lint against #[no_mangle]
Given that [`#[no_mangle]` is unsafe](https://github.com/rust-lang/rust/issues/28179) there should be a lint that is capable of detecting it (there was a suggestion that `unsafe_code` should also detect it, but it might make sense to be able to lint it individually too).

### PR DESCRIPTION
**Pull request by Nemo157**
fixes #72188 

r? @estebank

### CODE REVIEW

File: src/librustc_lint/builtin.rs
Comment (Nemo157) [inline src/librustc_lint/builtin.rs:289]: I believe these are the only item kinds on which `#[no_mangle]` makes a difference, it seems like you can stick in a lot of places where it's ignored. Should this check everywhere, or only locations where the attribute actually affects compilation?

**Comment by ollie27** [comment]
`#[export_name = "..."]` is unsafe for the same reason as `#[no_mangle]` so it should also be detected by this lint.

File: src/librustc_lint/builtin.rs
Comment (estebank) [reply src/librustc_lint/builtin.rs:299]: Only in places that actually affect the resulting binary. The others should receive warnings or errors about the non effect already. It is good practice to avoid emitting multiple errors for the same cause. If that already happens, we try to silence the less informative one(s).

**Comment by Dylan-DPC** [comment]
@estebank any updates on this?

File: src/librustc_lint/builtin.rs
Comment (estebank) [inline src/librustc_lint/builtin.rs:277]: Can we use `it.name.span` instead here and in the other 2 places?

File: src/librustc_lint/builtin.rs
Comment (estebank) [inline src/librustc_lint/builtin.rs:278]: Could we add a `note` to explain why the annotation is problematic and or a link to the docs?

**Review by estebank — APPROVED** [review]
Sorry for the delay. We'll need to do a crater run with this to make sure we don't cause major regressions in the wild as this is being added to an existing lint. Beyond these nitpicks, it looks good to me.

**Comment by Nemo157** [comment]
@estebank fixed the nitpicks. Are you able to trigger the crater run?

**Comment by Dylan-DPC** [comment]
@craterbot check

**Comment by danielhenrymantilla** [comment]
Could the lint span point to the attribute itself, rather than the identifier of the item being defined?

File: src/test/ui/lint/lint-unsafe-code.stderr
Comment (danielhenrymantilla) [inline src/test/ui/lint/lint-unsafe-code.stderr:4]: Could the lint span point to the attribute itself, rather than the identifier of the item being defined?

File: src/test/ui/lint/lint-unsafe-code.stderr
Comment (danielhenrymantilla) [inline src/test/ui/lint/lint-unsafe-code.stderr:5]: Could the lint span point to the attribute itself, rather than the identifier of the item being defined?

**Comment by Mark-Simulacrum** [comment]
@craterbot retry seems to have been spurious

**Comment by Mark-Simulacrum** [comment]
@craterbot retry

**Comment by Nemo157** [comment]
Surprisingly only one expected regression from a crate using a combination of `deny(unsafe_code)` and a `no_mangle` function. Looking at it it is (IMO) a problematic usage of `no_mangle` so I've opened an upstream issue for them (https://github.com/async-rs/async-log/issues/13).

File: src/test/ui/lint/lint-unsafe-code.stderr
Comment (Nemo157) [reply src/test/ui/lint/lint-unsafe-code.stderr:5]: Makes sense, I'll try to get this changed tomorrow.

**Comment by Nemo157** [comment]
Rebased with no conflicts over the big move 😅

**Comment by jyn514** [comment]
@Mark-Simulacrum suggested this should go through a lang FCP. @Nemo157, can you write up a summary of what changes this makes? Then I can ping T-lang about an FCP.

**Comment by scottmcm** [comment]
Looks like only one regression in crater: https://github.com/async-rs/async-log/blob/ac19dddb27a4624edb6eed49adf3ad4f9473d988/src/backtrace.rs#L9-L15

It's using the `backtrace` crate to find the caller from inside the trait: https://github.com/async-rs/async-log/blob/ac19dddb27a4624edb6eed49adf3ad4f9473d988/src/logger.rs#L73

And the crate has `forbid(unsafe_code)`: https://github.com/async-rs/async-log/blob/ac19dddb27a4624edb6eed49adf3ad4f9473d988/src/lib.rs#L106

(I wonder whether this could become `track_caller` instead.)

---

As I understand it, this is adding linting for `#[no_mangle]` and `#[export_name]` as part of `unsafe_code` <https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unsafe-code>.  That's an allow-by-default-lint, and the usual intent I see from `forbid`ding it is to advertise that there's no need to worry about unsoundness, so I support extending it to other potential unsoundness.

@rfcbot fcp merge

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

I'm nominated for lang team discussion next week to draw attention to this.

**Comment by pnkfelix** [comment]
@rfcbot reviewed

**Comment by Dylan-DPC** [comment]
@estebank this is waiting for your final review. Looks ready imo 😃

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+

I'm sniping this one. Sorry @estebank =)

**Pull request merged**
